File: airbyte-commons-temporal/src/test/java/io/airbyte/commons/temporal/ConnectionManageUtilsTest.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.commons.temporal;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.airbyte.commons.temporal.exception.DeletedWorkflowException;
import io.airbyte.commons.temporal.scheduling.ConnectionManagerWorkflow;
import io.airbyte.metrics.lib.MetricClient;
import io.temporal.client.BatchRequest;
import io.temporal.client.WorkflowOptions;
import java.util.UUID;
import org.junit.jupiter.api.Test;

class ConnectionManageUtilsTest {

  @Test
  void signalAndRepairIfNeceesaryWhenNoWorkflowWillCreate() throws DeletedWorkflowException {
    final var mWorkflow = mock(WorkflowClientWrapped.class);
    final var mMetric = mock(MetricClient.class);
    final var cid = UUID.randomUUID();

    when(mWorkflow.newWorkflowStub(any(), any(WorkflowOptions.class)))
        .thenReturn(mock(ConnectionManagerWorkflow.class));
    when(mWorkflow.newSignalWithStartRequest()).thenReturn(mock(BatchRequest.class));

    final var utils = new ConnectionManagerUtils(mWorkflow, mMetric);
    utils.signalWorkflowAndRepairIfNecessary(cid, (workflow) -> null);
    // Because we do not mock the getConnectionManagerWorkflow call, the underlying call throws an
    // exception
    // and the logic recreates it.
    verify(mWorkflow).signalWithStart(any());
  }

}

File: airbyte-commons-temporal/src/test/kotlin/io/airbyte/commons/temporal/ConnectionManageUtilsTest.kt
/*
 * Copyright (c) 2020-2026 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.commons.temporal

import io.airbyte.commons.temporal.scheduling.ConnectionManagerWorkflow
import io.airbyte.metrics.MetricClient
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import io.temporal.api.common.v1.WorkflowExecution
import io.temporal.client.BatchRequest
import io.temporal.client.WorkflowOptions
import io.temporal.workflow.Functions
import org.junit.jupiter.api.Test
import java.util.UUID

internal class ConnectionManageUtilsTest {
  @Test
  fun signalAndRepairIfNecessaryWhenNoWorkflowWillCreate() {
    val mWorkflow = mockk<WorkflowClientWrapped>()
    val mMetric = mockk<MetricClient>(relaxed = true)
    val cid = UUID.randomUUID()

    every {
      mWorkflow.newWorkflowStub(
        any<Class<ConnectionManagerWorkflow>>(),
        any<WorkflowOptions>(),
      )
    } returns mockk<ConnectionManagerWorkflow>(relaxed = true)
    every { mWorkflow.newSignalWithStartRequest() } returns mockk<BatchRequest>(relaxed = true)
    every { mWorkflow.signalWithStart(any<BatchRequest>()) } returns mockk<WorkflowExecution>(relaxed = true)

    val utils = ConnectionManagerUtils(mWorkflow, mMetric)
    utils.signalWorkflowAndRepairIfNecessary(cid) { _: ConnectionManagerWorkflow? -> Functions.Proc {} }
    // Because we do not mock the getConnectionManagerWorkflow call, the underlying call throws an
    // exception
    // and the logic recreates it.
    verify { mWorkflow.signalWithStart(any<BatchRequest>()) }
  }
}
